File: sdks/markets/ts/src/instructions/market/gpt.ts
/**
 * GPT Subject builders (Ix 15 CreateSubject, Ix 16 RequestAi).
 */
import { Address, TransactionInstruction } from "@solana/web3.js";

import { Ix, MARKET_PROGRAM_ID, SYSTEM_PROGRAM_ID } from "../../constants.js";
import * as pda from "../../pda.js";
import type { AddressInput } from "../../pda.js";
import { addr, pubkeyBytes, ro, u32LE, u64LE, u8, w, withDisc } from "../payload.js";

export interface CreateSubjectArgs {
  payer: AddressInput;
  nonce: bigint | number;
  optionsCount: number;
  llmContext: AddressInput;
  programId?: Address;
}

export async function createSubject(args: CreateSubjectArgs): Promise<TransactionInstruction> {
  const programId = args.programId ?? MARKET_PROGRAM_ID;
  const subject = await pda.subject(args.nonce, programId);
  return new TransactionInstruction({
    programId,
    keys: [w(addr(args.payer), true), w(subject.address), ro(SYSTEM_PROGRAM_ID)],
    data: withDisc(
      Ix.CreateSubject,
      u64LE(args.nonce),
      u8(args.optionsCount),
      pubkeyBytes(args.llmContext),
    ),
  });
}

export interface RequestAiArgs {
  subject: AddressInput;
  payer: AddressInput;
  text: string | Uint8Array;
  programId?: Address;
}

export async function requestAi(args: RequestAiArgs): Promise<TransactionInstruction> {
  const programId = args.programId ?? MARKET_PROGRAM_ID;
  const text = typeof args.text === "string" ? new TextEncoder().encode(args.text) : args.text;
  return new TransactionInstruction({
    programId,
    keys: [ro(addr(args.subject)), w(addr(args.payer), true), ro(SYSTEM_PROGRAM_ID)],
    data: withDisc(Ix.RequestAi, u32LE(text.length), text),
  });
}

File: sdks/markets/rust/src/ix/gpt.rs
//! GPT Subject builders (Ix 15 `CreateSubject`, Ix 16 `RequestAi`).

use crate::*;
use solana_sdk::instruction::{AccountMeta, Instruction};
use solana_sdk::pubkey::Pubkey;
use solana_sdk::system_program;

/// `CreateSubject` (Ix 15). Payload: `nonce u64 LE ++ options_count u8 ++ llm_context [32]`.
pub fn create_subject(
    payer: &Pubkey,
    nonce: u64,
    options_count: u8,
    llm_context: &Pubkey,
) -> Instruction {
    let (subject, _) = crate::pda::subject(nonce);
    let mut data = vec![IX_CREATE_SUBJECT];
    data.extend_from_slice(&nonce.to_le_bytes());
    data.push(options_count);
    data.extend_from_slice(llm_context.as_ref());
    Instruction {
        program_id: PROGRAM_ID,
        accounts: vec![
            AccountMeta::new(*payer, true),
            AccountMeta::new(subject, false),
            AccountMeta::new_readonly(system_program::id(), false),
        ],
        data,
    }
}

/// `RequestAi` (Ix 16) short form (no GPT CPI remaining accounts).
/// Payload: `text_len u32 LE ++ text`.
pub fn request_ai(subject: &Pubkey, payer: &Pubkey, text: &[u8]) -> Instruction {
    let mut data = vec![IX_REQUEST_AI];
    data.extend_from_slice(&(text.len() as u32).to_le_bytes());
    data.extend_from_slice(text);
    Instruction {
        program_id: PROGRAM_ID,
        accounts: vec![
            AccountMeta::new_readonly(*subject, false),
            AccountMeta::new(*payer, true),
            AccountMeta::new_readonly(system_program::id(), false),
        ],
        data,
    }
}
